platform/darwin/src/NSExpression+MGLAdditions.mm

```diff
@@ -27,37 +29,47 @@ @implementation NSExpression (MGLAdditions)
-            return { number.unsignedShortValue };
-        } else if (strcmp([number objCType], @encode(int)) == 0) {
-            return { number.intValue };
-        } else if (strcmp([number objCType], @encode(unsigned int)) == 0) {
-            return { number.unsignedIntValue };
-        } else if (strcmp([number objCType], @encode(long)) == 0) {
-            return { number.longValue };
-        } else if (strcmp([number objCType], @encode(unsigned long)) == 0) {
-            return { number.unsignedLongValue };
-        } else if (strcmp([number objCType], @encode(long long)) == 0) {
-            return { number.longLongValue };
-        } else if (strcmp([number objCType], @encode(unsigned long long)) == 0) {
-            return { number.unsignedLongLongValue };
-        } else if (strcmp([number objCType], @encode(float)) == 0) {
-            return { number.floatValue };
+        NSLog(@"NSNumber %@ (%s)", number, number.objCType);
+
+        // 32/64-bit check based on CGFloat storage size
+        // per https://developer.apple.com/library/ios/documentation/General/Conceptual/CocoaTouch64BitGuide/
+        BOOL thirtyTwoBit = sizeof(CGFloat) == sizeof(float);
+        BOOL sixtyFourBit = sizeof(CGFloat) == sizeof(double);
+        NSAssert((thirtyTwoBit || sixtyFourBit) && ! (thirtyTwoBit && sixtyFourBit), @"Fault in 32/64-bit determination");
+        NSLog(@"Running in %@-bit mode", (thirtyTwoBit ? @"32" : @"64"));
+
+        if ((strcmp(number.objCType, @encode(char)) == 0) ||
+            (strcmp(number.objCType, @encode(BOOL)) == 0)) {
+            // char: 32-bit boolean
+            // BOOL: 64-bit boolean
+            return { (bool)number.boolValue };
+        } else if (strcmp([number objCType], @encode(short))     == 0 ||
+                   strcmp([number objCType], @encode(int))       == 0 ||
+                   strcmp([number objCType], @encode(long))      == 0 ||
+                   strcmp([number objCType], @encode(long long)) == 0 ||
+                   strcmp([number objCType], @encode(NSInteger)) == 0) {
+            // In practice, all non-boolean whole numbers are converted to one
+            // of these types by NSExpression regardless of signedness.
+            return { (int64_t)number.integerValue };
+        } else if (strcmp([number objCType], @encode(unsigned char))      == 0 ||
+                   strcmp([number objCType], @encode(unsigned short))     == 0 ||
+                   strcmp([number objCType], @encode(unsigned int))       == 0 ||
+                   strcmp([number objCType], @encode(unsigned long))      == 0 ||
+                   strcmp([number objCType], @encode(unsigned long long)) == 0 ||
+                   strcmp([number objCType], @encode(NSUInteger))         == 0) {
+            // This code path does not seem to get run in current NSExpression
+            // implementations (even with e.g. [NSNumber numberWithLong:INT_MAX]),
```

Comment:
Comments updated in 18e84ca5e98fc73dc480a7b009492b98051716d8. Now I feel better about leaving this code path in. 😄